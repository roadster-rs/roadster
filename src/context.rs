use async_trait::async_trait;

use crate::config::app_config::AppConfig;

#[derive(Debug, Clone)]
pub struct AppContext {
    pub config: AppConfig,
}

impl AppContext {
    pub async fn new(config: AppConfig) -> anyhow::Result<Self> {
        let context = Self { config };
        Ok(context)
    }
}

// Define a custom trait because the standard From/Into traits don't support async
#[async_trait]
pub trait FromAppContextRef {
    async fn from_app_context_ref(ctx: &AppContext) -> anyhow::Result<Box<Self>>;
}

// Define a custom trait because the standard From/Into traits don't support async
#[async_trait]
pub trait GetAppContextFromRef {
    async fn get_app_context(&self) -> anyhow::Result<AppContext>;
}

// Implement for `AppContext` so consumers can just use `AppContext` for their state if they want
#[async_trait]
impl FromAppContextRef for AppContext {
    async fn from_app_context_ref(ctx: &AppContext) -> anyhow::Result<Box<Self>> {
        Ok(Box::new(ctx.clone()))
    }
}

// Implement for `AppContext` so consumers can just use `AppContext` for their state if they want
#[async_trait]
impl GetAppContextFromRef for AppContext {
    async fn get_app_context(&self) -> anyhow::Result<AppContext> {
        Ok(self.clone())
    }
}
