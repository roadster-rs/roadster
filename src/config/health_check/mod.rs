use crate::app::context::AppContext;
use crate::config::CustomConfig;
use crate::util::serde::default_true;
use axum_core::extract::FromRef;
use config::{FileFormat, FileSourceString};
use serde_derive::{Deserialize, Serialize};
use serde_with::serde_as;
use std::collections::BTreeMap;
use std::time::Duration;
use validator::Validate;

pub fn default_config() -> config::File<FileSourceString, FileFormat> {
    config::File::from_str(include_str!("default.toml"), FileFormat::Toml)
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct HealthCheck {
    #[serde(default = "default_true")]
    pub default_enable: bool,

    #[validate(nested)]
    pub max_duration: MaxDuration,

    #[cfg(feature = "db-sql")]
    #[validate(nested)]
    pub database: HealthCheckConfig<crate::config::EmptyConfig>,

    #[cfg(feature = "sidekiq")]
    #[validate(nested)]
    pub sidekiq: HealthCheckConfig<crate::config::EmptyConfig>,

    #[cfg(feature = "email-smtp")]
    #[validate(nested)]
    pub smtp: HealthCheckConfig<crate::config::EmptyConfig>,

    /// Allows providing configs for custom health checks. Any configs that aren't pre-defined above
    /// will be collected here.
    ///
    /// # Examples
    ///
    /// ```toml
    /// [health-check.foo]
    /// enable = true
    /// x = "y"
    /// ```
    ///
    /// This will be parsed as:
    /// ```raw
    /// HealthCheck#custom: {
    ///     "foo": {
    ///         HealthCheckConfig#common: {
    ///             enable: true,
    ///             priority: 10
    ///         },
    ///         HealthCheckConfig<CustomConfig>#custom: {
    ///             "x": "y"
    ///         }
    ///     }
    /// }
    /// ```
    #[serde(flatten)]
    #[validate(nested)]
    pub custom: BTreeMap<String, HealthCheckConfig<CustomConfig>>,
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct CommonConfig {
    // Optional so we can tell the difference between a consumer explicitly enabling/disabling
    // the health check, vs the health check being enabled/disabled by default.
    // If this is `None`, the value will match the value of `HealthCheck#default_enable`.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub enable: Option<bool>,
}

impl CommonConfig {
    pub fn enabled<S>(&self, state: &S) -> bool
    where
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
    {
        self.enable.unwrap_or(
            AppContext::from_ref(state)
                .config()
                .health_check
                .default_enable,
        )
    }
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct HealthCheckConfig<T: Validate> {
    #[serde(flatten)]
    #[validate(nested)]
    pub common: CommonConfig,
    #[serde(flatten)]
    #[validate(nested)]
    pub custom: T,
}

/// The maximum duration to wait for health checks to succeed before timing out and assuming
/// the checks failed.
#[serde_as]
#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct MaxDuration {
    /// The maximum time (in milliseconds) to wait  when running checks on app startup.
    #[serde_as(as = "serde_with::DurationMilliSeconds")]
    pub startup: Duration,
    /// The maximum time (in milliseconds) to wait when running checks via a web API (HTTP, gRPC, etc).
    /// In the default `_health` HTTP endpoint, this can be overridden via the `maxDuration`
    /// query parameter.
    #[serde_as(as = "serde_with::DurationMilliSeconds")]
    pub api: Duration,
    /// The maximum time (in milliseconds) to wait when running checks via the CLI. This can be
    /// overridden via the `-d/--max-duration` CLI arg.
    #[serde_as(as = "serde_with::DurationMilliSeconds")]
    pub cli: Duration,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::AppConfig;
    use rstest::rstest;

    #[rstest]
    #[case(true, None, true)]
    #[case(true, Some(true), true)]
    #[case(true, Some(false), false)]
    #[case(false, None, false)]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn common_config_enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.health_check.default_enable = default_enable;

        let context = AppContext::test(Some(config), None, None).unwrap();

        let common_config = CommonConfig { enable };

        // Act/Assert
        assert_eq!(common_config.enabled(&context), expected_enabled);
    }
}
