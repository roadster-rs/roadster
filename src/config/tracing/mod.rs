#[cfg(feature = "otel")]
use crate::util::serde_util::default_true;
use config::{FileFormat, FileSourceString};
use serde_derive::{Deserialize, Serialize};
#[cfg(feature = "otel")]
use url::Url;
use validator::Validate;

pub fn default_config() -> config::File<FileSourceString, FileFormat> {
    config::File::from_str(include_str!("default.toml"), FileFormat::Toml)
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct Tracing {
    pub level: String,

    /// The name of the service to use for the OpenTelemetry `service.name` field. If not provided,
    /// will use the [`App::name`][crate::config::app_config::App] config value, translated to `snake_case`.
    #[cfg(feature = "otel")]
    pub service_name: Option<String>,

    /// Propagate traces across service boundaries. Mostly useful in microservice architectures.
    #[serde(default = "default_true")]
    #[cfg(feature = "otel")]
    pub trace_propagation: bool,

    /// URI of the OTLP exporter where traces/metrics/logs will be sent.
    #[cfg(feature = "otel")]
    pub otlp_endpoint: Option<Url>,
}

// To simplify testing, these are only run when all of the config fields are available
#[cfg(all(test, feature = "otel"))]
mod deserialize_tests {
    use super::*;
    use crate::util::test_util::TestCase;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(
        r#"
        level = "debug"
        "#
    )]
    #[case(
        r#"
        level = "info"
        service-name = "foo"
        "#
    )]
    #[case(
        r#"
        level = "error"
        trace-propagation = false
        "#
    )]
    #[case(
        r#"
        level = "debug"
        otlp-endpoint = "https://example.com:1234"
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn sidekiq(_case: TestCase, #[case] config: &str) {
        let tracing: Tracing = toml::from_str(config).unwrap();

        assert_toml_snapshot!(tracing);
    }
}
