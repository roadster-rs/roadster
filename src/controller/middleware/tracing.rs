use crate::app_context::AppContext;
use crate::controller::middleware::Middleware;
use axum::extract::MatchedPath;
use axum::http::{Request, Response};
use axum::Router;
use opentelemetry_semantic_conventions::trace::{
    HTTP_REQUEST_METHOD, HTTP_RESPONSE_STATUS_CODE, HTTP_ROUTE, URL_PATH,
};
use serde_derive::{Deserialize, Serialize};
use std::time::Duration;
use tower_http::trace::{DefaultOnResponse, MakeSpan, OnRequest, OnResponse, TraceLayer};
use tracing::{event, field, info_span, Level, Span, Value};

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct TracingConfig {}

pub struct TracingMiddleware;
impl Middleware for TracingMiddleware {
    fn name(&self) -> String {
        "tracing".to_string()
    }

    fn enabled(&self, context: &AppContext) -> bool {
        context.config.middleware.tracing.common.enabled(context)
    }

    fn priority(&self, context: &AppContext) -> i32 {
        context.config.middleware.tracing.common.priority
    }

    fn install(&self, router: Router, _context: &AppContext) -> anyhow::Result<Router> {
        let request_id_header_name = &_context
            .config
            .middleware
            .set_request_id
            .custom
            .common
            .header_name;

        let router = router.layer(
            TraceLayer::new_for_http()
                .make_span_with(CustomMakeSpan::new(request_id_header_name.clone()))
                .on_request(CustomOnRequest)
                .on_response(CustomOnResponse::new()),
        );

        Ok(router)
    }
}

#[derive(Debug, Clone)]
pub struct CustomMakeSpan {
    pub request_id_header_name: String,
}

impl CustomMakeSpan {
    pub fn new(request_id_header_name: String) -> Self {
        Self {
            request_id_header_name,
        }
    }
}

impl<B> MakeSpan<B> for CustomMakeSpan {
    fn make_span(&mut self, request: &Request<B>) -> Span {
        let path = get_path(request);
        let request_id = get_request_id(&self.request_id_header_name, request);
        info_span!("request",
            { HTTP_REQUEST_METHOD } = %request.method(),
            { HTTP_ROUTE } = optional_trace_field(path),
            request_id = optional_trace_field(request_id),
            // Fields that aren't know at request time, but will (may?) be known by
            // response time
            { HTTP_RESPONSE_STATUS_CODE } = field::Empty,
        )
    }
}

fn get_path<B>(request: &Request<B>) -> Option<&str> {
    request
        .extensions()
        .get::<MatchedPath>()
        .map(|path| path.as_str())
}

fn get_request_id<'a, B>(
    request_id_header_name: &'a str,
    request: &'a Request<B>,
) -> Option<&'a str> {
    request
        .headers()
        .get(request_id_header_name)
        .and_then(|v| v.to_str().ok())
}

fn optional_trace_field<T>(value: Option<T>) -> Box<dyn Value>
where
    T: ToString,
{
    value
        .map(|x| Box::new(field::display(x.to_string())) as Box<dyn Value>)
        .unwrap_or(Box::new(field::Empty))
}

#[derive(Debug, Copy, Clone)]
pub struct CustomOnRequest;

impl<B> OnRequest<B> for CustomOnRequest {
    fn on_request(&mut self, request: &Request<B>, _: &Span) {
        event!(
            Level::INFO,
            version = ?request.version(),
            { URL_PATH } = %request.uri(),
            request_headers = ?request.headers(),
            "started processing request",
        )
    }
}

#[derive(Debug, Clone)]
pub struct CustomOnResponse {
    default: DefaultOnResponse,
}

impl CustomOnResponse {
    pub fn new() -> CustomOnResponse {
        CustomOnResponse {
            default: DefaultOnResponse::new()
                .include_headers(true)
                // TODO: Configure the level via AppConfig?
                .level(Level::INFO),
        }
    }
}

impl Default for CustomOnResponse {
    fn default() -> Self {
        Self::new()
    }
}

impl<B> OnResponse<B> for CustomOnResponse {
    fn on_response(self, response: &Response<B>, latency: Duration, span: &Span) {
        span.record(HTTP_RESPONSE_STATUS_CODE, response.status().as_u16());
        self.default.on_response(response, latency, span);
    }
}
