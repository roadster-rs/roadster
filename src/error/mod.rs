pub mod api;
pub mod auth;
#[cfg(feature = "http")]
pub mod axum;
#[cfg(feature = "cli")]
pub mod cli;
pub mod config;
#[cfg(feature = "db-sql")]
pub mod db;
#[cfg(feature = "email")]
pub mod email;
#[cfg(feature = "http")]
pub mod mime;
pub mod mutex;
pub mod other;
pub mod parse;
#[cfg(feature = "worker-pg")]
pub mod pgmq;
pub mod reqwest;
pub mod serde;
#[cfg(feature = "worker-sidekiq")]
pub mod sidekiq;
pub mod tokio;
#[cfg(feature = "grpc")]
pub mod tonic;
pub mod tracing;
#[cfg(feature = "worker")]
pub mod worker;

use crate::app::context::extension::ExtensionRegistryError;
use crate::error::api::ApiError;
use crate::error::auth::AuthError;
#[cfg(feature = "http")]
use crate::error::axum::AxumError;
#[cfg(feature = "cli")]
use crate::error::cli::CliError;
#[cfg(feature = "db-sql")]
use crate::error::db::DbError;
#[cfg(feature = "email")]
use crate::error::email::EmailError;
#[cfg(feature = "http")]
use crate::error::mime::MimeError;
use crate::error::mutex::MutexError;
use crate::error::other::OtherError;
use crate::error::parse::ParseError;
#[cfg(feature = "worker-pg")]
use crate::error::pgmq::PgmqError;
use crate::error::reqwest::ReqwestError;
use crate::error::serde::SerdeError;
#[cfg(feature = "worker-sidekiq")]
use crate::error::sidekiq::SidekiqError;
use crate::error::tokio::TokioError;
#[cfg(feature = "grpc")]
use crate::error::tonic::TonicError;
use crate::error::tracing::TracingError;
#[cfg(feature = "worker")]
use crate::error::worker::WorkerError;
use crate::health::check::registry::HealthCheckRegistryError;
use crate::lifecycle::registry::LifecycleHandlerRegistryError;
use crate::service::registry::ServiceRegistryError;
#[cfg(feature = "http")]
use ::axum::http::StatusCode;
#[cfg(feature = "http")]
use ::axum::response::{IntoResponse, Response};
#[cfg(feature = "open-api")]
use aide::OperationOutput;
use std::convert::Infallible;
use thiserror::Error;

pub type RoadsterResult<T> = Result<T, Error>;

#[derive(Debug, Error)]
#[non_exhaustive]
pub enum Error {
    #[error(transparent)]
    Api(#[from] ApiError),

    #[error(transparent)]
    Auth(#[from] AuthError),

    #[error(transparent)]
    Serde(#[from] SerdeError),

    #[cfg(feature = "db-sql")]
    #[error(transparent)]
    Db(#[from] DbError),

    #[cfg(feature = "worker-sidekiq")]
    #[error(transparent)]
    Sidekiq(#[from] SidekiqError),

    #[cfg(feature = "worker-pg")]
    #[error(transparent)]
    Pgmq(#[from] PgmqError),

    #[cfg(feature = "worker")]
    #[error(transparent)]
    Worker(#[from] WorkerError),

    #[cfg(feature = "cli")]
    #[error(transparent)]
    Clap(#[from] clap::error::Error),

    #[error(transparent)]
    Config(#[from] config::ConfigError),

    #[error(transparent)]
    Validation(#[from] validator::ValidationErrors),

    #[error(transparent)]
    Io(#[from] std::io::Error),

    #[error(transparent)]
    Tokio(#[from] TokioError),

    #[error(transparent)]
    Tracing(#[from] TracingError),

    #[error(transparent)]
    Reqwest(#[from] ReqwestError),

    #[cfg(feature = "http")]
    #[error(transparent)]
    Axum(#[from] AxumError),

    #[error(transparent)]
    Parse(#[from] ParseError),

    #[cfg(feature = "http")]
    #[error(transparent)]
    Mime(#[from] MimeError),

    #[cfg(feature = "grpc")]
    #[error(transparent)]
    Tonic(#[from] TonicError),

    #[cfg(feature = "email")]
    #[error(transparent)]
    Email(#[from] EmailError),

    #[error(transparent)]
    HealthCheckRegistry(#[from] HealthCheckRegistryError),

    #[error(transparent)]
    LifecycleHandlerRegistry(#[from] LifecycleHandlerRegistryError),

    #[error(transparent)]
    ServiceRegistry(#[from] ServiceRegistryError),

    #[error(transparent)]
    ExtensionRegistry(#[from] ExtensionRegistryError),

    #[error(transparent)]
    Mutex(#[from] MutexError),

    #[error(transparent)]
    Infallible(#[from] Infallible),

    #[cfg(feature = "test-containers")]
    #[error(transparent)]
    TestContainers(
        #[from] testcontainers_modules::testcontainers::core::error::TestcontainersError,
    ),

    #[cfg(feature = "cli")]
    #[error(transparent)]
    Cli(#[from] CliError),

    #[error(transparent)]
    Other(#[from] OtherError),
}

#[cfg(feature = "http")]
impl IntoResponse for Error {
    fn into_response(self) -> Response {
        ::tracing::debug!("{}", self);
        match self {
            Error::Api(err) => err.into_response(),
            Error::Auth(_) => StatusCode::UNAUTHORIZED.into_response(),
            _ => StatusCode::INTERNAL_SERVER_ERROR.into_response(),
        }
    }
}

#[cfg(feature = "open-api")]
impl OperationOutput for Error {
    type Inner = api::http::HttpError;
}

#[cfg(test)]
mod tests {
    #[cfg(feature = "http")]
    use axum_core::response::IntoResponse;

    #[rstest::fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> crate::testing::snapshot::TestCase {
        Default::default()
    }

    #[rstest::rstest]
    #[case(
        crate::error::api::http::HttpError::bad_request().to_err()
    )]
    #[case(
        crate::error::api::http::HttpError::unauthorized().to_err()
    )]
    #[case(
        crate::error::api::http::HttpError::forbidden().to_err()
    )]
    #[case(
        crate::error::api::http::HttpError::not_found().to_err()
    )]
    #[case(
        crate::error::api::http::HttpError::gone().to_err()
    )]
    #[case(
        crate::error::api::http::HttpError::internal_server_error().to_err()
    )]
    #[case(
        crate::error::api::http::HttpError::from(axum::http::StatusCode::BAD_REQUEST).into()
    )]
    #[case(
        axum::http::StatusCode::BAD_REQUEST.into()
    )]
    #[case(
        crate::error::api::ApiError::Other(Box::new(crate::error::other::OtherError::Message("error".to_owned()))).into()
    )]
    #[cfg(feature = "http")]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn into_response(
        _case: crate::testing::snapshot::TestCase,
        #[case] error: crate::error::Error,
    ) {
        insta::assert_debug_snapshot!(error.into_response().status());
    }
}
