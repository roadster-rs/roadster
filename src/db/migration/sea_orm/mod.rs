use crate::app::context::AppContext;
use crate::db::migration::{DownArgs, MigrationInfo, MigrationStatus, Migrator, UpArgs};
use crate::error::RoadsterResult;
use async_trait::async_trait;
use axum_core::extract::FromRef;
use sea_orm_migration::MigratorTrait;
use std::cmp::min;
use std::marker::PhantomData;

pub mod check;
pub mod collation;
pub mod schema;
pub mod timestamp;
pub mod user;
pub mod uuid;

impl From<sea_orm_migration::Migration> for MigrationInfo {
    fn from(value: sea_orm_migration::Migration) -> Self {
        Self {
            name: value.name().to_string(),
            status: value.status().into(),
        }
    }
}

impl From<sea_orm_migration::MigrationStatus> for MigrationStatus {
    fn from(value: sea_orm_migration::MigrationStatus) -> Self {
        match value {
            sea_orm_migration::MigrationStatus::Applied => Self::Applied,
            sea_orm_migration::MigrationStatus::Pending => Self::Pending,
        }
    }
}

pub struct SeaOrmMigrator<M>
where
    M: MigratorTrait + Send + Sync,
{
    migrator: PhantomData<M>,
}

impl<M> SeaOrmMigrator<M>
where
    M: MigratorTrait + Send + Sync,
{
    pub fn new(_migrator: M) -> Self {
        Self {
            migrator: Default::default(),
        }
    }
}

#[async_trait]
impl<S, M> Migrator<S> for SeaOrmMigrator<M>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    M: MigratorTrait + Send + Sync,
{
    #[tracing::instrument(skip_all)]
    async fn up(&self, state: &S, args: &UpArgs) -> crate::error::RoadsterResult<usize> {
        let context = crate::app::context::AppContext::from_ref(state);
        let pending = M::get_pending_migrations(context.sea_orm()).await?;

        let to_run = if let Some(steps) = args.steps {
            min(steps, pending.len())
        } else {
            pending.len()
        };

        M::up(context.sea_orm(), args.steps.map(|steps| steps as u32)).await?;

        // Assume all of the pending steps (up to `args.steps` count) ran successfully.
        Ok(to_run)
    }

    #[tracing::instrument(skip_all)]
    async fn down(&self, state: &S, args: &DownArgs) -> RoadsterResult<usize> {
        let context = crate::app::context::AppContext::from_ref(state);
        let applied = M::get_applied_migrations(context.sea_orm()).await?;

        let to_roll_back = if let Some(steps) = args.steps {
            min(steps, applied.len())
        } else {
            applied.len()
        };

        M::down(context.sea_orm(), args.steps.map(|steps| steps as u32)).await?;

        // Assume all of the applied steps (up to `args.steps` count) were rolled back successfully.
        Ok(to_roll_back)
    }

    #[tracing::instrument(skip_all)]
    async fn status(&self, state: &S) -> RoadsterResult<Vec<MigrationInfo>> {
        let context = crate::app::context::AppContext::from_ref(state);

        let migrations = M::get_migration_with_status(context.sea_orm())
            .await?
            .into_iter()
            .map(|migration| migration.into())
            .collect();

        Ok(migrations)
    }
}
