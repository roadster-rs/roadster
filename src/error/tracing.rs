use crate::error::Error;

#[derive(Debug, Error)]
#[non_exhaustive]
pub enum TracingError {
    /// An error that occurs during tracing initialization.
    #[error(transparent)]
    Init(#[from] TracingInitError),

    #[error(transparent)]
    #[cfg(any(feature = "worker-pg", feature = "db-sea-orm"))]
    ParseLevel(#[from] log::ParseLevelError),

    #[error(transparent)]
    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
}

#[derive(Debug, Error)]
#[non_exhaustive]
pub enum TracingInitError {
    #[cfg(feature = "otel")]
    #[error(transparent)]
    OtelTrace(#[from] opentelemetry_sdk::trace::TraceError),

    #[cfg(feature = "otel")]
    #[error(transparent)]
    ExporterBuilder(#[from] opentelemetry_otlp::ExporterBuildError),

    #[error(transparent)]
    ParseLevel(#[from] tracing::metadata::ParseLevelError),

    #[error(transparent)]
    ParseFilter(#[from] tracing_subscriber::filter::ParseError),

    #[error(transparent)]
    FilterFromEnv(#[from] tracing_subscriber::filter::FromEnvError),

    #[error(transparent)]
    Init(#[from] tracing_subscriber::util::TryInitError),
}

#[cfg(feature = "otel")]
impl From<opentelemetry_sdk::trace::TraceError> for Error {
    fn from(value: opentelemetry_sdk::trace::TraceError) -> Self {
        Self::Tracing(TracingError::from(TracingInitError::from(value)))
    }
}

#[cfg(feature = "otel")]
impl From<opentelemetry_otlp::ExporterBuildError> for Error {
    fn from(value: opentelemetry_otlp::ExporterBuildError) -> Self {
        Self::Tracing(TracingError::from(TracingInitError::from(value)))
    }
}

impl From<tracing::metadata::ParseLevelError> for Error {
    fn from(value: tracing::metadata::ParseLevelError) -> Self {
        Self::Tracing(TracingError::from(TracingInitError::from(value)))
    }
}

#[cfg(any(feature = "worker-pg", feature = "db-sea-orm"))]
impl From<log::ParseLevelError> for Error {
    fn from(value: log::ParseLevelError) -> Self {
        Self::Tracing(TracingError::from(value))
    }
}

impl From<tracing_subscriber::filter::ParseError> for Error {
    fn from(value: tracing_subscriber::filter::ParseError) -> Self {
        Self::Tracing(TracingError::from(TracingInitError::from(value)))
    }
}

impl From<tracing_subscriber::filter::FromEnvError> for Error {
    fn from(value: tracing_subscriber::filter::FromEnvError) -> Self {
        Self::Tracing(TracingError::from(TracingInitError::from(value)))
    }
}

impl From<tracing_subscriber::util::TryInitError> for Error {
    fn from(value: tracing_subscriber::util::TryInitError) -> Self {
        Self::Tracing(TracingError::from(TracingInitError::from(value)))
    }
}
