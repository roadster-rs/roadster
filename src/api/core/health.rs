use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::health_check::{CheckResponse, ErrorData, HealthCheck, Status};
#[cfg(feature = "open-api")]
use aide::OperationIo;
#[cfg(any(feature = "sidekiq", feature = "email-smtp"))]
use anyhow::anyhow;
use axum_core::extract::FromRef;
use futures::future::join_all;
#[cfg(feature = "open-api")]
use schemars::JsonSchema;
#[cfg(feature = "db-sql")]
use sea_orm::DatabaseConnection;
use serde_derive::{Deserialize, Serialize};
use serde_with::{serde_as, skip_serializing_none};
#[cfg(feature = "sidekiq")]
use sidekiq::redis_rs::cmd;
use std::collections::BTreeMap;
use std::sync::Arc;
use std::time::Duration;
use std::time::Instant;
use tokio::time::timeout;
use tracing::{debug, error, info, instrument};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "open-api", derive(JsonSchema, OperationIo))]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct HeathCheckResponse {
    /// Total latency of checking the health of the app.
    pub latency: u128,
    pub resources: BTreeMap<String, CheckResponse>,
}

#[instrument(skip_all)]
pub async fn health_check<S>(
    state: &S,
    duration: Option<Duration>,
) -> RoadsterResult<HeathCheckResponse>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    if let Some(duration) = duration.as_ref() {
        info!(
            "Running checks for a maximum duration of {} ms",
            duration.as_millis()
        );
    } else {
        info!("Running checks");
    }
    let context = AppContext::from_ref(state);
    let timer = Instant::now();

    let check_futures = context.health_checks().into_iter().map(|check| {
        Box::pin(async move {
            let name = check.name();
            info!(%name, "Running check");
            let check_timer = Instant::now();
            let result = match run_check(check, duration).await {
                Ok(response) => {
                    info!(%name, latency_ms=%response.latency, "Check completed");
                    match &response.status {
                        Status::Ok => {}
                        Status::Err(_) => {
                            error!(%name, "Resource is not healthy");
                            debug!(%name, "Error details: {response:?}");
                        }
                    }
                    response
                }
                Err(err) => CheckResponse::builder()
                    .status(Status::Err(
                        ErrorData::builder()
                            .msg(format!(
                                "An error occurred while running health check `{name}`: {err}"
                            ))
                            .build(),
                    ))
                    .latency(check_timer.elapsed())
                    .build(),
            };
            (name, result)
        })
    });

    let resources = join_all(check_futures).await.into_iter().collect();

    let latency = timer.elapsed().as_millis();

    info!(latency_ms=%latency, "Checks completed");

    Ok(HeathCheckResponse { latency, resources })
}

async fn run_check(
    check: Arc<dyn HealthCheck>,
    duration: Option<Duration>,
) -> RoadsterResult<CheckResponse> {
    if let Some(duration) = duration {
        timeout(duration, check.check()).await?
    } else {
        check.check().await
    }
}

#[serde_as]
#[skip_serializing_none]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "open-api", derive(JsonSchema))]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct Latency {
    /// How long it takes to acquire a connection from the pool.
    pub acquire_conn_latency: Option<u128>,
    /// How long it takes to ping the resource after the connection is acquired.
    pub ping_latency: Option<u128>,
}

#[cfg(feature = "db-sql")]
pub(crate) async fn db_health(context: &AppContext, duration: Option<Duration>) -> CheckResponse {
    let db_timer = Instant::now();
    let db_status = match ping_db(context.db(), duration).await {
        Ok(_) => Status::Ok,
        Err(err) => Status::Err(ErrorData::builder().msg(err.to_string()).build()),
    };
    let db_timer = db_timer.elapsed();
    CheckResponse::builder()
        .status(db_status)
        .latency(db_timer)
        .build()
}

#[cfg(feature = "db-sql")]
#[instrument(skip_all)]
async fn ping_db(db: &DatabaseConnection, duration: Option<Duration>) -> RoadsterResult<()> {
    if let Some(duration) = duration {
        timeout(duration, db.ping()).await??;
    } else {
        db.ping().await?;
    }
    Ok(())
}

#[cfg(feature = "email-smtp")]
pub(crate) async fn smtp_health(context: &AppContext, duration: Option<Duration>) -> CheckResponse {
    let timer = Instant::now();
    let status = match ping_smtp(context.mailer(), duration).await {
        Ok(_) => Status::Ok,
        Err(err) => Status::Err(ErrorData::builder().msg(err.to_string()).build()),
    };
    let timer = timer.elapsed();
    CheckResponse::builder()
        .status(status)
        .latency(timer)
        .build()
}

#[cfg(feature = "email-smtp")]
async fn ping_smtp(
    mailer: &lettre::SmtpTransport,
    duration: Option<Duration>,
) -> RoadsterResult<()> {
    let connected = if let Some(duration) = duration {
        timeout(duration, async { mailer.test_connection() }).await??
    } else {
        mailer.test_connection()?
    };
    if connected {
        Ok(())
    } else {
        Err(anyhow!("Not connected to the SMTP server").into())
    }
}

#[cfg(feature = "sidekiq")]
#[instrument(skip_all)]
pub(crate) async fn redis_health(
    redis: &sidekiq::RedisPool,
    duration: Option<Duration>,
) -> CheckResponse {
    let redis_timer = Instant::now();
    let (redis_status, acquire_conn_latency, ping_latency) = match ping_redis(redis, duration).await
    {
        Ok((a, b)) => (Status::Ok, Some(a.as_millis()), Some(b.as_millis())),
        Err(err) => (
            Status::Err(ErrorData::builder().msg(err.to_string()).build()),
            None,
            None,
        ),
    };
    let redis_timer = redis_timer.elapsed();
    CheckResponse::builder()
        .status(redis_status)
        .latency(redis_timer)
        .custom(Latency {
            acquire_conn_latency,
            ping_latency,
        })
        .build()
}

#[cfg(feature = "sidekiq")]
#[instrument(skip_all)]
async fn ping_redis(
    redis: &sidekiq::RedisPool,
    duration: Option<Duration>,
) -> RoadsterResult<(Duration, Duration)> {
    let timer = Instant::now();
    let mut conn = if let Some(duration) = duration {
        timeout(duration, redis.get()).await??
    } else {
        redis.get().await?
    };
    let acquire_conn_latency = timer.elapsed();

    let timer = Instant::now();
    let msg = uuid::Uuid::new_v4().to_string();
    let pong: String = cmd("PING")
        .arg(&msg)
        .query_async(conn.unnamespaced_borrow_mut())
        .await?;
    let ping_latency = timer.elapsed();

    if pong == msg {
        Ok((acquire_conn_latency, ping_latency))
    } else {
        Err(anyhow!("Ping response does not match input.").into())
    }
}
