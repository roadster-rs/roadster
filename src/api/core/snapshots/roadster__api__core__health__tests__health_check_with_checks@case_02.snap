---
source: src/api/core/health.rs
expression: health_response
---
{
  "latency": 1,
  "resources": {
    "example": {
      "status": {
        "err": {
          "msg": "Error"
        }
      },
      "latency": 200
    }
  }
}
